src/Stratis.Features.Diagnostic/PeerDiagnostic/PeerStatisticsCollector.cs

```diff
@@ -0,0 +1,168 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Stratis.Bitcoin.AsyncWork;
+using Stratis.Bitcoin.EventBus;
+using Stratis.Bitcoin.Signals;
+using Stratis.Bitcoin.Utilities;
+using Stratis.Features.Diagnostic.Controllers;
+using Event = Stratis.Bitcoin.EventBus.CoreEvents;
+
+namespace Stratis.Features.Diagnostic.PeerDiagnostic
+{
+    /// <summary>
+    /// Subscribe to peer events and keep track of their activities.
+    /// A summary of peer activities can be obtained using <see cref="DiagnosticController"/> actions
+    /// </summary>
+    public sealed class PeerStatisticsCollector : IDisposable
+    {
+        private object lockStartStopCollecting;
+
+        private readonly IAsyncProvider asyncProvider;
+        private readonly ISignals signals;
+        private readonly INodeLifetime nodeLifetime;
+        private readonly DiagnosticSettings diagnosticSettings;
+
+        /// <summary>Track current collecting status, when true Peer Collector is collecting statistics.</summary>
+        public bool Enabled { get; private set; }
+
+        private readonly Dictionary<IPEndPoint, PeerStatistics> peersStatistics;
+
+        /// <summary>Non blocking queue that consume received peer events to generate peer statistics.</summary>
+        private IAsyncDelegateDequeuer<Event.PeerEventBase> peersEventsQueue;
+
+        /// <summary>Holds a list of event subscriptions.</summary>
+        private readonly List<SubscriptionToken> eventSubscriptions;
+
+        public PeerStatisticsCollector(IAsyncProvider asyncProvider, ISignals signals, DiagnosticSettings diagnosticSettings, INodeLifetime nodeLifetime)
+        {
+            this.asyncProvider = asyncProvider;
+            this.signals = Guard.NotNull(signals, nameof(signals));
+            this.nodeLifetime = nodeLifetime;
+            this.diagnosticSettings = Guard.NotNull(diagnosticSettings, nameof(diagnosticSettings));
+
+            this.eventSubscriptions = new List<SubscriptionToken>();
+            this.peersStatistics = new Dictionary<IPEndPoint, PeerStatistics>();
+
+            this.lockStartStopCollecting = new object();
+        }
+
+        public void Initialize()
+        {
+            this.Enabled = this.diagnosticSettings.PeersStatisticsCollectorEnabled;
+            if (this.Enabled)
+                StartCollecting();
```

Comment:
this.

Reply:
good finding.
If no more important things to fix about this PR, I'll fix this in another one following the merge (builds are failing randomly so having green builds is something to care about :D)